elki-outlier/src/main/java/elki/outlier/isolation/IsolationForest.java

```diff
@@ -0,0 +1,366 @@
+      for(int i = 0; i < numTrees; i++) {
+        ModifiableDBIDs sample = DBIDUtil.randomSample(relation.getDBIDs(), subsampleSize, rnd);
+        trees.add(new Tree( //
+            sample, //
+            random, //
+            0, //
+            (int) FastMath.ceil(FastMath.log(relation.size()) / FastMath.log(2)), //
+            relation));
+        LOG.incrementProcessed(prog);
+      }
+      LOG.ensureCompleted(prog);
+
+      // Iterate over all objects
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance()) {
+        final double score = distributionForInstance(relation.get(iter));
+        scores.putDouble(iter, score);
+        minmax.put(score);
+        LOG.incrementProcessed(prog);
+      }
+      LOG.ensureCompleted(prog);
+    }
+
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("iforest anomaly score", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  /**
+   * Returns the average path length of an unsuccessful search. Returns 0 if
+   * argument is less than or equal to 1.
+   */
+  protected static double c(double n) {
+    if(n <= 1.0) {
+      return 0;
+    }
+    return 2 * (FastMath.log(n - 1) + 0.5772156649) - (2 * (n - 1) / n);
+  }
+
+  /**
+   * Returns distribution of scores.
+   */
+  // protected double[] distributionForInstance(V vector) {
+  protected double distributionForInstance(NumberVector vector) {
+    double avgPathLength = 0;
+    for(Tree m_tree : trees) {
+      avgPathLength += m_tree.pathLength(vector);
+    }
+    avgPathLength /= trees.size();
+
+    return FastMath.pow(2, -avgPathLength / c(subsampleSize));
+
+    // return scores;
+  }
+
+  /**
+   * Inner class for building and using an isolation tree.
+   */
+  static class Tree {
+
+    /**
+     * The size of the node
+     */
+    private int size;
+
+    /**
+     * The split attribute
+     */
+    private int a;
+
+    /**
+     * The split point
+     */
+    private double q;
+
+    /**
+     * The successors
+     */
+    private List<Tree> successors;
+
+    /**
+     * Constructs a tree from data.
+     *
+     * @param data Subsample from data
+     * @param r Random generator
+     * @param height
+     * @param maxHeight
+     */
+    protected Tree(ModifiableDBIDs data, Random r, int height, int maxHeight, Relation<? extends NumberVector> relation) {
+      // Set size of node
+      size = data.size();
+
+      // Stop splitting if necessary
+      if((size <= 1) || (height == maxHeight)) {
+        return;
+      }
+
+      // Compute mins and maxs and eligible attributes
+      final int dim = RelationUtil.dimensionality(relation);
```

Comment:
While this is O(1) and still fairly cheap, it involves a few type checks. It likely is best to pass along the dimensionality as additional parameter.